apps/emqx_durable_storage/src/emqx_ds_beamformer.erl

```diff
@@ -0,0 +1,828 @@
+    WaitingTab = emqx_ds_beamformer_waitq:new(),
+    S = #s{
+        module = CBM,
+        shard = ShardId,
+        metrics_id = shard_metrics_id(ShardId),
+        name = Name,
+        pending_queue = PendingTab,
+        wait_queue = WaitingTab,
+        pending_request_limit = cfg_pending_request_limit(),
+        batch_size = cfg_batch_size()
+    },
+    self() ! ?housekeeping_loop,
+    {ok, S}.
+
+handle_call(
+    Req = #poll_req{},
+    _From,
+    S = #s{pending_queue = PendingTab, wait_queue = WaitingTab, metrics_id = Metrics}
+) ->
+    NQueued = ets:info(PendingTab, size) + ets:info(WaitingTab, size),
+    case NQueued >= S#s.pending_request_limit of
+        true ->
+            emqx_ds_builtin_metrics:inc_poll_requests_dropped(Metrics, 1),
+            Reply = {error, recoverable, too_many_requests},
+            {reply, Reply, S};
+        false ->
+            ets:insert(S#s.pending_queue, Req),
+            {reply, ok, start_fulfill_loop(S)}
+    end;
+handle_call(_Call, _From, S) ->
+    {reply, {error, unknown_call}, S}.
+
+handle_cast(_Cast, S) ->
+    {noreply, S}.
+
+handle_info(#shard_event{events = Events}, S) ->
+    ?tp(debug, emqx_ds_beamformer_event, #{events => Events}),
+    {noreply, maybe_fulfill_waiting(S, Events)};
+handle_info(?fulfill_loop, S0) ->
+    S1 = S0#s{is_spinning = false},
+    S = fulfill_pending(S1),
+    {noreply, S};
+handle_info(?housekeeping_loop, S0) ->
+    %% Reload configuration according from environment variables:
+    S1 = S0#s{
+        batch_size = cfg_batch_size(),
+        pending_request_limit = cfg_pending_request_limit()
+    },
+    S = cleanup(S1),
+    erlang:send_after(cfg_housekeeping_interval(), self(), ?housekeeping_loop),
+    {noreply, S};
+handle_info(_Info, S) ->
+    {noreply, S}.
+
+terminate(_Reason, #s{shard = ShardId, name = Name}) ->
+    Pool = emqx_ds_beamformer_sup:pool(ShardId),
+    gproc_pool:disconnect_worker(Pool, Name),
+    gproc_pool:remove_worker(Pool, Name),
+    ok.
+
+%%================================================================================
+%% Internal exports
+%%================================================================================
+
+-spec start_link(module(), _Shard, integer(), opts()) -> {ok, pid()}.
+start_link(Mod, ShardId, Name, Opts) ->
+    gen_server:start_link(?MODULE, [Mod, ShardId, Name, Opts], []).
+
+%% @doc RPC target: split the beam and dispatch replies to local
+%% consumers.
+-spec do_dispatch(beam()) -> ok.
+do_dispatch(Beam = #beam{}) ->
+    lists:foreach(
+        fun({{Alias, ItKey}, Result}) ->
+            Alias ! #poll_reply{ref = Alias, userdata = ItKey, payload = Result}
+        end,
+        split(Beam)
+    ).
+
+%%================================================================================
+%% Internal functions
+%%================================================================================
+
+-spec start_fulfill_loop(s()) -> s().
+%% @TODO start_fulfill_loop is missleading
```

Comment:
Why is it misleading and how it can be renamed?

Reply:
To me it just queue a message/signal in the mailbox, this is just 'scheduled', I wont call it "start_x_loop" which sounds like do it now.
Just mislead me a bit when read in the caller's context.

Reply:
Logically it's a loop, it just designed to avoid blocking the `gen_server`.